File: week2_algorithmic_warmup/2_last_digit_of_fibonacci_number/fibonacci_last_digit.py
import sys
sys.set_int_max_str_digits(0)

def fibonacci_last_digit(n):     
    if n <= 1:
        return n

    previous = 0
    current  = 1

    for _ in range(2, n + 1):
        previous, current = current, previous + current

    
    return current % 10

def fibo_last_digit(n):
    if n <= 9:
        return n
    
    return n % 10


if __name__ == '__main__':
    n = int(input())
    print(fibonacci_last_digit(n))

File: week2_algorithmic_warmup/2_last_digit_of_fibonacci_number/fibonacci_last_digit.cpp
#include <iostream>

int get_fibonacci_last_digit_naive(int n) {
    if (n <= 1)
        return n;

    long long previous = 0;
    long long current  = 1;
    long long tmp_previous = 0;

    for (int i = 0; i < n - 1; ++i) {
        tmp_previous = previous;
        previous = current;
        current = (tmp_previous + current) % 10;
    }

    return current;
}

int main() {
    int n;
    std::cin >> n;
    int c = get_fibonacci_last_digit_naive(n);
    std::cout << c << '\n';
    }
